### PR TITLE
Update features to fork solution once instead of forking once per changed document.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/CSharp/Portable/CodeRefactorings/EnableNullable/EnableNullableCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -67,14 +68,21 @@ private static bool ShouldOfferRefactoring(Project project)
...
-        foreach (var document in project.Documents)
-        {
-            if (await document.IsGeneratedCodeAsync(cancellationToken).ConfigureAwait(false))
-                continue;
+        var updatedDocumentRoots = await ProducerConsumer<(DocumentId documentId, SyntaxNode newRoot, PreservationMode mode)>.RunParallelAsync(
+            source: project.Documents,
+            produceItems: static async (document, callback, fallbackOptions, cancellationToken) =>
+            {
+                if (await document.IsGeneratedCodeAsync(cancellationToken).ConfigureAwait(false))
+                    return;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/EnableNullable/EnableNullableCodeRefactoringProvider.cs:76]: view with whitespace off.

File: src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs [hunk 2]
```diff
@@ -773,40 +767,39 @@ await FixReferencesAsync(document, changeNamespaceService, addImportService, ref
...
-            }
 
             var document = solution.GetRequiredDocument(id);
             linkedDocumentsToSkip.AddRange(document.GetLinkedDocumentIds());
-            documentsToProcessBuilder.Add(document);
-
-            document = await RemoveUnnecessaryImportsWorkerAsync(
-                document,
-                CreateImports(document, names, withFormatterAnnotation: false),
-                cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs:793]: this was wacky.  we were updating the documents serially, and then doing a parallel processing of it below.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 3]
```diff
@@ -1553,25 +1539,17 @@ public SolutionCompilationState RemoveAnalyzerConfigDocuments(ImmutableArray<Doc
             addDocumentsToProjectState);
     }
 
-    private SolutionCompilationState AddDocumentsToMultipleProjects<TDocumentState>(
-        IEnumerable<(ProjectId projectId, ImmutableArray<TDocumentState> newDocumentStates)> projectIdAndNewDocuments,
-        Func<ProjectState, ImmutableArray<TDocumentState>, TranslationAction> addDocumentsToProjectState)
-        where TDocumentState : TextDocumentState
-    {
-        return UpdateDocumentsInMultipleProjects(projectIdAndNewDocuments, addDocumentsToProjectState);
-    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1562]: removed this

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 4]
```diff
@@ -714,6 +714,19 @@ public SolutionCompilationState WithDocumentText(DocumentId documentId, SourceTe
...
+    {
+        return WithDocumentContents(
+            texts, IsUnchanged,
+            static (documentState, text, mode) => documentState.UpdateText(text, mode));
+
+        static bool IsUnchanged(DocumentState oldDocument, SourceText text)
+            => oldDocument.TryGetText(out var oldText) && text == oldText;
+    }
+
+    private SolutionCompilationState WithDocumentContents<TContent>(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:726]: added this helper used by WithDocumentTexts and WithDocumentSyntaxRoots

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Features/Core/Portable/IntroduceParameter/AbstractIntroduceParameterCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -241,17 +242,26 @@ CodeAction CreateNewCodeAction(string actionName, bool allOccurrences, Introduce
         var rewriter = new IntroduceParameterDocumentRewriter(this, originalDocument,
             expression, methodSymbol, containingMethod, selectedCodeAction, fallbackOptions, allOccurrences);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/IntroduceParameter/AbstractIntroduceParameterCodeRefactoringProvider.cs:243]: i checked, and rewriter should be safe to use concurrently from within produceItems below.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski For review when you get back.

**Pull request merged**